controllers/topology/current_state.go

```
@@ -18,12 +18,140 @@ package topology
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	infra, err := r.getTemplate(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries for all MachineDeployments and filters them for their linked Cluster and
+// whether they are managed by a ClusterClass using labels. A Cluster may have zero or more MachineDeployments. Zero is
+// expected on first reconcile. If MachineDeployments are found for the Cluster their Infrastructure and Bootstrap references
+// are inspected. Where these are not found the function will throw an error.
+func (r *ClusterReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *v1alpha4.Cluster) (map[string]machineDeploymentTopologyState, error) {
+	mDeploymentState := make(map[string]machineDeploymentTopologyState)
+
+	md := &v1alpha4.MachineDeploymentList{}
+	err := r.Client.List(ctx, md, client.MatchingLabels{
+		v1alpha4.ClusterLabelName:         cluster.Name,
+		v1alpha4.ClusterTopologyLabelName: "",
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read MachineDeployments for managed topology")
+	}
+	for _, m := range md.Items {
+		machineDeployment := m
+		if _, ok := m.ObjectMeta.Labels[v1alpha4.ClusterTopologyMachineDeploymentLabelName]; !ok {
+			return nil, fmt.Errorf("failed to find MachineDeploymentTopology name for %s", machineDeployment.Name)
+		}
+		mdTopologyName := m.ObjectMeta.Labels[v1alpha4.ClusterTopologyMachineDeploymentLabelName]
+		if mdTopologyName == "" {
+			return nil, fmt.Errorf("no value for MachineDeploymentTopoplogyName for %s", machineDeployment.Name)
```

```suggestion
			return nil, fmt.Errorf("no value for the %s label in %s", v1alpha4.ClusterTopologyMachineDeploymentLabelName, machineDeployment.Name)
```

---

Also, let's raise an error if we find two machineDeployments with the same topology name (that can be checked by verifying the topology name does not exists in the map)